megatron/model/rwkv/rwkv.py

```diff
@@ -0,0 +1,356 @@
+        xxx = torch.bmm(xxx, self.time_maa_w2).view(5, B, T, -1)
+        mw, mk, mv, mr, mg = xxx.unbind(dim=0)
+
+        xw = x + xx * (self.time_maa_w + mw)
+        xk = x + xx * (self.time_maa_k + mk)
+        xv = x + xx * (self.time_maa_v + mv)
+        xr = x + xx * (self.time_maa_r + mr)
+        xg = x + xx * (self.time_maa_g + mg)
+
+        r = self.receptance(xr)
+        k = self.key(xk)
+        v = self.value(xv)
+        g = F.silu(self.gate(xg))
+
+        ww = torch.tanh(xw @ self.time_decay_w1) @ self.time_decay_w2
+        w = self.time_decay + ww
+
+        return r, k, v, g, w
+
+    def jit_func_2(self, x, g):
+        B, T, C = x.size()
+        x = x.view(B * T, C)
+
+        x = self.ln_x(x).view(B, T, C)
+        x = self.output(x * g)
+        return x
+
+    def forward(self, x):
+        B, T, C = x.size()
+        H = self.neox_args.num_attention_heads
+
+        r, k, v, g, w = self.jit_func(x)
+        x = RUN_CUDA_RWKV(B, T, C, H, r, k, v, w, u=self.time_faaaa)
+
+        return self.jit_func_2(x, g)
+
+
+class RWKV_ChannelMix(nn.Module):
+    """
+    Channel Mix layer. The ffn in RWKV
+    """
+
+    def __init__(self, neox_args, layer_number):
+        super().__init__()
+        self.neox_args = neox_args
+        self.layer_number = layer_number
+        self.time_shift = nn.ZeroPad2d((0, 0, 1, -1))
+
+        with torch.no_grad():  # fancy init of time_mix
+            ratio_1_to_almost0 = 1.0 - (layer_number / neox_args.num_layers)  # 1 to ~0
+            ddd = torch.ones(1, 1, neox_args.hidden_size)
+            for i in range(neox_args.hidden_size):
+                ddd[0, 0, i] = i / neox_args.hidden_size
+            self.time_maa_k = nn.Parameter(1.0 - torch.pow(ddd, ratio_1_to_almost0))
+            self.time_maa_r = nn.Parameter(1.0 - torch.pow(ddd, ratio_1_to_almost0))
+
+        self.key = nn.Linear(neox_args.hidden_size, neox_args.dim_ffn, bias=False)
+        self.receptance = nn.Linear(
+            neox_args.hidden_size, neox_args.hidden_size, bias=False
+        )
+        self.value = nn.Linear(neox_args.dim_ffn, neox_args.hidden_size, bias=False)
+
+    def forward(self, x):
+        xx = self.time_shift(x) - x
+        xk = x + xx * self.time_maa_k
+        xr = x + xx * self.time_maa_r
+
+        k = self.key(xk)
+        k = torch.relu(k) ** 2
+        kv = self.value(k)
+        return torch.sigmoid(self.receptance(xr)) * kv
+
+
+class RWKVResidualLayer(nn.Module):
+    """
+    RWKV layer definition
+    """
+
+    def __init__(self, neox_args, layer_number):
+        super().__init__()
+        self.neox_args = neox_args
+        self.layer_number = layer_number
+        self.fp16 = neox_args.precision == "fp16"
+        self.bf16 = neox_args.precision == "bfloat16"
+        if not hasattr(neox_args, "dim_att"):
+            neox_args.dim_att = neox_args.hidden_size
+        if not hasattr(neox_args, "dim_ffn"):
```

Comment:
RWKV 5+ use a 3.5x FFN multiplier, not 4x. There is unfortunately some misleading code in the RWKV-LM repo that has led many to make this mistake.

Reply:
I think we should probably let people experiment with this ratio however they want though, right? We can provide scripts that train the official versions, but if someone wants to train with a different ratio let them (much like we do with transformers).

Reply:
Oh yeah it definitely needs to stay parameterizable by the end user. The only change I'm requesting is that the default be 3.5x - this is done to reduce the parameters taken up by the gate that was added in v5, to stay at equiparams, and is what is in every released RWKV-6.0 model and described in the paper.

As an aside, it looks like v6.1+ is going to remove the gate and this will go back up to a default of 4x at that time.

Reply:
Added versioning dirnames and changed this to 3.5x for v6